internal/resolver/delegatingresolver/delegatingresolver.go

```diff
@@ -0,0 +1,311 @@
+
+	proxyTarget := resolver.Target{URL: *r.proxyURL}
+	wcc := &wrappingClientConn{
+		stateListener: r.updateProxyResolverState,
+		parent:        r}
+	return proxyBuilder.Build(proxyTarget, wcc, opts)
+}
+
+func (r *delegatingResolver) ResolveNow(o resolver.ResolveNowOptions) {
+	if r.targetResolver != nil {
+		r.targetResolver.ResolveNow(o)
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.ResolveNow(o)
+	}
+}
+
+func (r *delegatingResolver) Close() {
+	if r.targetResolver != nil {
+		r.targetResolver.Close()
+		r.targetResolver = nil
+	}
+	if r.proxyResolver != nil {
+		r.proxyResolver.Close()
+		r.proxyResolver = nil
+	}
+}
+
+// combinedAddressesLocked creates a list of combined addresses by
+// pairing each proxy address with every target address. For each pair, it
+// generates a new [resolver.Address] using the proxy address, and adding the
+// target address as the attribute along with user info.
+func (r *delegatingResolver) combinedAddressesLocked() ([]resolver.Address, []resolver.Endpoint) {
+	var addresses []resolver.Address
+	for _, proxyAddr := range r.proxyAddrs {
+		for _, targetAddr := range r.targetAddrs {
+			newAddr := resolver.Address{Addr: proxyAddr.Addr}
+			newAddr = proxyattributes.Populate(newAddr, proxyattributes.Options{
+				User:        r.proxyURL.User,
+				ConnectAddr: targetAddr.Addr,
+			})
+			addresses = append(addresses, newAddr)
+		}
+	}
+
+	// Create a list of combined addresses by pairing each proxy endpoint
+	// with every target endpoint. For each pair, it constructs a new
+	// [resolver.Endpoint] using the first address from the proxy endpoint,
+	// as the DNS resolver is expected to return only one address per proxy
+	// endpoint. The target address and user information from the proxy URL
+	// are added as attributes to this address. The resulting list of addresses
+	// is then grouped into endpoints, covering all combinations of proxy and
+	// target endpoints.
+	var endpoints []resolver.Endpoint
+	for _, proxyEndpt := range r.proxyEndpoint {
+		proxyAddr := proxyEndpt.Addresses[0].Addr
+		for _, endpt := range r.targetEndpoint {
+			var addrs []resolver.Address
+			for _, targetAddr := range endpt.Addresses {
+				newAddr := resolver.Address{Addr: proxyAddr}
+				newAddr = proxyattributes.Populate(newAddr, proxyattributes.Options{
+					User:        r.proxyURL.User,
+					ConnectAddr: targetAddr.Addr,
+				})
+				addrs = append(addrs, newAddr)
+			}
+			endpoints = append(endpoints, resolver.Endpoint{Addresses: addrs})
+		}
+	}
+	return addresses, endpoints
+}
+
+// updateProxyResolverState updates the proxy resolver state by storing proxy
+// addresses and endpoints, marking the resolver as ready, and triggering a
+// state update if both proxy and target resolvers are ready. It is a
+// StateListener function of wrappingClientConn passed to the proxy resolver.
+func (r *delegatingResolver) updateProxyResolverState(state resolver.State) error {
```

Comment:
That's a good point. I think blocking until both resolvers have produced at least one update seems like an OK thing to do, because the delegating resolver anyways does not send an update to the channel until that happens. 

Another option would be:
- continue to return `nil` from `updateProxyResolverState` and `updateTargetResolverState`, if both resolvers haven't yet produced an update. If both resolvers have already produced at least one update, these methods should return the value returned from `r.cc.UpdateState`.
- And from `updateStateIfReadyLocked` when `UpdateState` is called for the first time, if it returns an error, then we could call `ResolveNow` on both child resolvers.

But I'm not sure if the above approach is worth the complexity that it brings.

Reply:
Done. Used the second approach as discussed with Easwar offline.

Reply:
It would be good to capture the information about why the blocking approach does not work (somewhere in a comment).